### PR TITLE
Test polars support

### PR DESCRIPTION
**Pull request by TheooJ**
**What is this PR addressing?**
Testing polars inputs (closing issue #825)

**Method**
-  I didn't duplicate tests but rather used `@pytest.mark.parametrize` with the two modules.
- If Polars isn’t available, functions are only testing for Pandas dataframes as discussed in https://github.com/skrub-data/skrub/pull/785
- When both modules have an `assert_frame_equal` testing function, I’ve created lists of tuples to store both methods and call the needed one in the testing functions.

**Example**
```
from pandas.testing import assert_frame_equal
from skrub._dataframe._polars import POLARS_SETUP

MODULES = [pd]  # This is used when we don't need assert_frame_equal
ASSERT_TUPLES = [(pd, assert_frame_equal)]

if POLARS_SETUP:
    import polars as pl
    from polars.testing import assert_frame_equal as assert_frame_equal_pl

    MODULES.append(pl)
    ASSERT_TUPLES.append((pl, assert_frame_equal_pl))

# This was modified to either return a pd or pl DataFrame
def get_mixed_datetime_format(px, as_array=False):
    """px should either be the Pandas or Polars module."""
    df = px.DataFrame(
        dict(
            a=[
                "2022-10-15",
                "2021-12-25",
                "2020-05-18",
                "2019-10-15 12:00:00",
            ]
        )
    )
    if as_array:
        return df.to_numpy()
    return df

# When available, this is testing both pd and pl modules
@pytest.mark.parametrize("px, assert_frame_equal_", ASSERT_TUPLES)
def test_indempotency(px, assert_frame_equal_):
    df = get_mixed_datetime_format(px)
    df_dt = to_datetime(df)
    df_dt_2 = to_datetime(df_dt)
    assert_frame_equal_(df_dt, df_dt_2)

    X_trans = DatetimeEncoder().fit_transform(df)
    X_trans_2 = DatetimeEncoder().fit_transform(df_dt)
    assert_array_equal(X_trans, X_trans_2)
```
------- 
**Comments/Discussion**
- Some test functions specified that they returned `→ None`, some didn’t. I removed all occurrences of these in the tests I’ve covered
- Some functions created dataframes using `pd.DataFrame([[1, 3], [2, 4]], columns=['a', 'b'])`. This method isn’t supported by Polars, so I’ve replaced it by `pd.DataFrame({'col1': [1, 2], 'col2': [3, 4]})` when applicable
- Some dataframes that are being tested used `pd.NA`. I used `pl.Null` for the Polars equivalent
- When the test failed in polars, i put 
    ```
    def test_working_for_pd_but_not_pl(px):
        if px is pl:
            pytest.xfail(reason="Some useful message for debugging later")
    ```
- The errors most commonly raised are that `pl.DataFrames` don’t have a `reset_index()` method, and that `pl.DataFrame.drop()` doesn't have an `axis` argument

### CODE REVIEW

File: skrub/tests/test_interpolation_join.py [hunk 1]
```diff
@@ -27,11 +35,23 @@ def weather():
     )
 
 
+@pytest.mark.parametrize("px", MODULES)
 @pytest.mark.parametrize("key", [["latitude", "longitude"], "latitude"])
 @pytest.mark.parametrize("with_nulls", [False, True])
-def test_interpolation_join(buildings, weather, key, with_nulls):
+def test_interpolation_join(px, buildings, weather, key, with_nulls):
```
Comment (Vincent-Maladiere) [inline skrub/tests/test_interpolation_join.py:41]: You have errors because you need to "shield" the call to `pl` with `if POLARS_SETUP`. This could be done with:
```python
if POLARS_SETUP and px is pl:
```
(note that when the first element is False, the second element won't be evaluated)

File: skrub/tests/test_similarity_encoder.py [hunk 2]
```diff
@@ -322,10 +339,14 @@ def test_check_fitted_super_vectorizer() -> None:
...
-def test_inverse_transform():
+@pytest.mark.parametrize("px", MODULES)
+def test_inverse_transform(px):
     encoder = SimilarityEncoder()
-    encoder.set_output(transform="pandas")
-    X = pd.DataFrame({"A": ["aaa", "aax", "xxx"], "B": ["bbb", "bby", "yyy"]})
+    if px is pd:
+        encoder.set_output(transform="pandas")
+    elif px is pl:
+        encoder.set_output(transform="polars")
```
Comment (Vincent-Maladiere) [inline skrub/tests/test_similarity_encoder.py:348]: The issue is that this is not possible ([yet](https://github.com/scikit-learn/scikit-learn/pull/27315)!). I guess we can only `xfail` that one.

File: skrub/tests/test_datetime_encoder.py [hunk 3]
```diff
@@ -62,22 +78,24 @@ def get_nanoseconds(as_array=False):
     return df
 
 
-def get_nan_datetime(as_array=False):
-    df = pd.DataFrame(
+def get_nan_datetime(px, as_array=False):
+    """px should either be the Pandas or Polars module."""
+    df = px.DataFrame(
```
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_datetime_encoder.py:83]: I agree, it avoids the issue with the NULL list, and the different `fillna()` methods names. Would you rather see it in the `get_dataframe(...)` functions (that would then take a `px` argument), or directly in the test function:
```
df = get_nan_datetime()
df = px.Dataframe(df)
```

File: skrub/tests/test_deduplicate.py [hunk 4]
```diff
@@ -47,7 +61,7 @@ def test_deduplicate(
     assert recovered_categories.shape[0] == n_clusters
     assert np.isin(clean_categories, recovered_categories).all()
     deduplicated_data = deduplicate(data, n_clusters=n_clusters)
-    translation_table = pd.Series(deduplicated_data, index=data)
+    translation_table = px.Series(deduplicated_data, index=data)
```
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_deduplicate.py:64]: This is solved with the same suggestion as above

```
translation_table = pd.Series(deduplicated_data, index=data)
translation_table = px.Series(translation_table)
```

File: skrub/tests/test_gap_encoder.py [hunk 5]
```diff
@@ -5,9 +5,19 @@
 from sklearn.model_selection import train_test_split
 
 from skrub import GapEncoder
+from skrub._dataframe._polars import POLARS_SETUP
 from skrub.datasets import fetch_midwest_survey
 from skrub.tests.utils import generate_data
 
+MODULES = [pd]
+NULL = {"pandas": pd.NA}
```
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_gap_encoder.py:13]: Removing all NULL dicts

File: skrub/tests/test_interpolation_join.py [hunk 6]
```diff
@@ -27,11 +35,23 @@ def weather():
...
+        pytest.xfail(
+            reason=(
+                "In polars, DataFrame.drop() got an unexpected keyword argument 'axis'"
+            )
+        )
+        weather = pl.DataFrame(weather)
+        buildings = pl.DataFrame(buildings)
     if not with_nulls:
-        weather = weather.fillna(0.0)
+        if px is pd:
```
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_interpolation_join.py:51]: Much better I agree, done

File: skrub/tests/test_fuzzy_join.py [hunk 7]
```diff
@@ -329,37 +366,39 @@ def test_numerical_column() -> None:
...
+@pytest.mark.parametrize("px, assert_frame_equal_", ASSERT_TUPLES)
+def test_datetime_column(px, assert_frame_equal_):
     """
     Testing that fuzzy_join works with datetime columns.
     """
-
-    left = pd.DataFrame(
+    if px is pl:
+        pytest.xfail(reason="Module 'polars' has no attribute 'to_datetime'")
+    left = px.DataFrame(
```
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_fuzzy_join.py:376]: So always use 

```
df = pd.DataFrame({"A": [0, 1], "B": [0, 1]})
df = px.DataFrame(df)
```

instead of ```df = px.DataFrame({"A": [0, 1], "B": [0, 1]})``` ?

I guess for this example it's the same, but the first formulation generalizes to nulls, missing values filling, and other specificities.

File: skrub/tests/test_datetime_encoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_datetime_encoder.py:83]: Note that when px.Dataframe isn't called, we don't test some functions with `pl.Null`. Ex:

```
observations = [
    ["alice", "bob"],
    [pd.NA, "alice"],
    ["bob", None],
    ["alice", "charlie"],
    [np.nan, "alice"],
]
observations = np.array(observations, dtype=object)
```

File: skrub/tests/test_fuzzy_join.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_fuzzy_join.py:376]: Similar example here:

```
# Numpy array with two columns
X = np.array([["alice", "charlie"], ["bob", "delta"]])
enc = GapEncoder(n_components=2, random_state=42)
X_enc_array = enc.fit_transform(X)
# Dataframe with two columns
df = px.DataFrame(X)
```

where `df = px.DataFrame(X)` could be `df = pd.DataFrame(X); df = px.DataFrame(df)`

File: skrub/tests/test_interpolation_join.py (hunk 1 above)
Comment (TheooJ) [reply skrub/tests/test_interpolation_join.py:41]: Solved with your above suggestion of calling `pd.DataFrame()` first

File: skrub/tests/test_datetime_encoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Vincent-Maladiere) [reply skrub/tests/test_datetime_encoder.py:83]: As discussed IRL, within the test function is fine

File: skrub/tests/test_deduplicate.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vincent-Maladiere) [reply skrub/tests/test_deduplicate.py:64]: Why wouldn't this work?

[the remark this answers is not in the document]
Comment (Vincent-Maladiere) [reply skrub/tests/test_deduplicate.py:64]: As discussed IRL, when having a parametrized Polars gets tricky, let's have a polar-specific function to run the main functions (here `deduplicate`). We'll revamp the specifics later.

File: skrub/tests/test_fuzzy_join.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Vincent-Maladiere) [reply skrub/tests/test_fuzzy_join.py:376]: Let's be flexible. If the dataframe is as trivial as `df = px.DataFrame({"A": [0, 1], "B": [0, 1]})`, using `px` directly is fine.

File: skrub/tests/test_datetime_encoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_datetime_encoder.py:83]: Done

File: skrub/tests/test_datetime_encoder.py [hunk 8]
```diff
@@ -143,7 +165,9 @@ def test_fit(
     add_day_of_the_week,
     resolution,
 ):
-    X = get_data_func(as_array=as_array)
+    if px is pl:
+        pytest.xfail(reason="Polars is not supported yet")
```
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_datetime_encoder.py]: This is solved creating pd.DataFrames first, then using `df = px.DataFrame(df)`. Now, all tests except for `test_mixed_type_dataframe` pass, and this one fails because of `X = to_datetime(X)` raises an error in Polars.

I assume it wasn't working when creating `px.DataFrames` directly because:

```
pl.DataFrame([
        ["2020-01-01", "2020-01-02", "2020-01-03"],
        ["2021-02-03", "2020-02-04", "2021-02-05"],
        ["2022-01-01", "2020-12-25", "2022-01-03"],
        ["2023-02-03", "2020-02-04", "2023-02-05"],
    ],
)
```
Is a 3 by 4 dataframe but

```
df = pd.DataFrame(
    [
        ["2020-01-01", "2020-01-02", "2020-01-03"],
        ["2021-02-03", "2020-02-04", "2021-02-05"],
        ["2022-01-01", "2020-12-25", "2022-01-03"],
        ["2023-02-03", "2020-02-04", "2023-02-05"],
    ],
)

pl.DataFrame(df)
```

is a 4 by 3 dataframe

File: skrub/tests/test_deduplicate.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TheooJ) [reply skrub/tests/test_deduplicate.py]: As discussed IRL, since this one isn't directly dependent on pandas or polas, I'm leaving it as is for now

File: skrub/tests/test_interpolation_join.py [hunk 9]
```diff
@@ -100,41 +141,76 @@ def test_condition_choice():
         ).fit(None)
 
 
-def test_suffix():
+@pytest.mark.parametrize("px", MODULES)
+def test_suffix(px):
+    if px is pl:
```
Comment (Vincent-Maladiere) [inline skrub/tests/test_interpolation_join.py:146]: You still have errors because of this

**Comment by jeromedockes** [comment]
out of curiosity in `is_namespace_polars` why is it necessary to do those checks rather than using `__name__`?

**Comment by Vincent-Maladiere** [comment]
To stay consistent with what we already have (e.g., `is_polars`), mostly? Kind of a nitpick, I agree. Also, this is something we might reuse. But let's revert if you prefer.

**Comment by jeromedockes** [comment]
my question was not about using a named function vs inlining it, but rather about checking `__name__` vs checking if polars is in sys.modules and then importing it and comparing with `is`

I was just interested to know if there is some situation you have encountered where comparing `__name__ == 'polars'` would not yield the expected result

**Comment by Vincent-Maladiere** [comment]
I saw similar patterns in scikit-learn to avoid weird behaviors like duck-typing. Happy to reverse if this brings unnecessary complexity, though.

**Comment by jeromedockes** [comment]
> I saw similar patterns in scikit-learn to avoid weird behaviors like duck-typing. Happy to reverse if this brings unnecessary complexity, though.

I see something similar in utils.validation._is_pandas_df but in that case they need to check if the input is a pandas dataframe rather than if a module is pandas so it is a bit different.
btw in _is_pandas_df they read the module directly from sys.modules rather than re-importing

I don't really have a preference here I'm just trying to learn in which situations I shouldn't rely on modules' __name__ attribute

(as ATM we just use for testing though, I would be in favor of putting it in a private module like testing_utils and starting with the simplest implementation. In any case +1 for your suggestion to put it in a function  so we can easily change it in one place only)

**Comment by Vincent-Maladiere** [comment]
>  I would be in favor of putting it in a private module like testing_utils and starting with the simplest implementation

Let's apply this advice using `px.__name__ == "polars"`, @TheooJ, then we can merge.

**Comment by TheooJ** [comment]
I'm rewriting the function as 
```
def is_namespace_polars(px):
    return px.__name__ == "polars"
```

in `tests/_utils.py`

**Review by Vincent-Maladiere — APPROVED** [review]
LGTM, thank you @TheooJ !

**Comment by jeromedockes** [comment]
maybe `testing_utils.py` so it doesn't get picked up by the [test discovery](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#conventions-for-python-test-discovery) rules

**Comment by jeromedockes** [comment]
> this is also used in this [exact form in the array API](https://github.com/data-apis/array-api-compat/blob/main/array_api_compat/common/_helpers.py)

AFAICT this is also for checking a dataframe rather than a module directly?

**Comment by TheooJ** [comment]
> maybe `testing_utils.py` so it doesn't get picked up by the [test discovery](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#conventions-for-python-test-discovery) rules

I think it discovers `test_*.py` or `*_test.py` files but not `_test_*.py` files

**Comment by jeromedockes** [comment]
> I think it discovers `test_*.py` or `*_test.py` files but not `_test_*.py` files

yes sorry I read too fast and missed the '_'!

File: skrub/_dataframe/_test_utils.py [hunk 10]
```diff
@@ -0,0 +1,6 @@
+def is_namespace_pandas(px):
```
Comment (jeromedockes) [inline skrub/_dataframe/_test_utils.py:1]: I think "module" would be more precise than "namespace"

File: skrub/tests/test_fuzzy_join.py [hunk 11]
```diff
@@ -9,19 +9,33 @@
...
 
+MODULES = [pd]
+ASSERT_TUPLES = [(pd, assert_frame_equal)]
 
+if POLARS_SETUP:
+    import polars as pl
+    from polars.testing import assert_frame_equal as assert_frame_equal_pl
+
+    MODULES.append(pl)
+    ASSERT_TUPLES.append((pl, assert_frame_equal_pl))
```
Comment (jeromedockes) [inline skrub/tests/test_fuzzy_join.py:23]: let's leave it for another pr but I think we should reorganize this a bit to put this kind of setup either in the `_test_utils.py` or in `conftest.py` and avoid duplicating it for each test module

**Review by jeromedockes — APPROVED** [review]
LGTM, thanks a lot @TheooJ ! it's super useful. and all those xfails show us what still needs to be done for polars support :)

File: skrub/_dataframe/_test_utils.py (hunk 10 above)
Comment (TheooJ) [reply skrub/_dataframe/_test_utils.py:1]: I agree, done

**Comment by jeromedockes** [comment]
@TheooJ sorry about that but we now have a small git merge conflict. could you fix that and then I will merge it

**Comment by jeromedockes** [comment]
thanks a lot!

**Comment by Vincent-Maladiere** [comment]
> AFAICT this is also for checking a dataframe rather than a module directly?

To check a module directly, they use `sys.module` as well:
https://github.com/data-apis/array-api-compat/blob/51d7832f048f58d47237d35fb1e7561d8ebfe259/array_api_compat/common/_helpers.py#L119
